tools/reaction_analyzer/include/subscriber.hpp

```diff
@@ -0,0 +1,250 @@
+  std::unique_ptr<message_filters::Subscriber<PublishedTime>> sub2_;
+  std::unique_ptr<message_filters::Synchronizer<ExactTimePolicy>> synchronizer_;
+  // tmp: only for the messages who don't have header e.g. AckermannControlCommand
+  std::unique_ptr<message_filters::Cache<PublishedTime>> cache_;
+};
+
+// Variant to create subscribers for different message types
+using SubscriberVariablesVariant = std::variant<
+  SubscriberVariables<PointCloud2>, SubscriberVariables<DetectedObjects>,
+  SubscriberVariables<TrackedObjects>, SubscriberVariables<PredictedObjects>,
+  SubscriberVariables<Trajectory>, SubscriberVariables<AckermannControlCommand>>;
+
+// The supported message types
+enum class SubscriberMessageType {
+  UNKNOWN = 0,
+  ACKERMANN_CONTROL_COMMAND = 1,
+  TRAJECTORY = 2,
+  POINTCLOUD2 = 3,
+  DETECTED_OBJECTS = 4,
+  PREDICTED_OBJECTS = 5,
+  TRACKED_OBJECTS = 6,
+};
+
+// Reaction Types
+enum class ReactionType {
+  UNKNOWN = 0,
+  FIRST_BRAKE = 1,
+  SEARCH_ZERO_VEL = 2,
+  SEARCH_ENTITY = 3,
+};
+
+// The configuration of the topic to be subscribed which are defined in reaction_chain
+struct TopicConfig
+{
+  std::string node_name;
+  std::string topic_address;
+  std::string time_debug_topic_address;
+  SubscriberMessageType message_type;
+};
+
+// Place for the reaction functions' parameter configuration
+struct FirstBrakeParams
+{
+  bool debug_control_commands;
+  double control_cmd_buffer_time_interval;
+  size_t min_number_descending_order_control_cmd;
+  double min_jerk_for_brake_cmd;
+};
+
+struct SearchZeroVelParams
+{
+  double max_looking_distance;
+};
+
+struct SearchEntityParams
+{
+  double search_radius;
+};
+
+// Place for the store the reaction parameter configuration (currently only for first brake)
+struct ReactionParams
+{
+  FirstBrakeParams first_brake_params;
+  SearchZeroVelParams search_zero_vel_params;
+  SearchEntityParams search_entity_params;
+};
+
+using ChainModules = std::vector<TopicConfig>;
+
+class SubscriberBase
+{
+public:
+  explicit SubscriberBase(
+    rclcpp::Node * node, Odometry::ConstSharedPtr & odometry, geometry_msgs::msg::Pose entity_pose,
+    std::atomic<bool> & spawn_object_cmd);
+
+  ~SubscriberBase() = default;
+
+  std::optional<std::unordered_map<std::string, BufferVariant>> getMessageBuffersMap();
+  void reset();
+
+private:
+  std::mutex mutex_;
+
+  rclcpp::Node * node_;
+  Odometry::ConstSharedPtr odometry_;
+  geometry_msgs::msg::Pose entity_pose_;
+  std::atomic<bool> & spawn_object_cmd_;
+
+  // Variables to be initialized in constructor
+  ChainModules chain_modules_;
+  ReactionParams reaction_params_{};
+
+  // Variants
+  std::unordered_map<std::string, SubscriberVariablesVariant> subscriber_variables_map_;
+  std::unordered_map<std::string, BufferVariant> message_buffers_;
+
+  // tf
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+
+  // Functions
+  void initReactionChainsAndParams();
```

Comment:
```suggestion
  void initReactionChainsAndParams();
```

Could you refactor all function names to snake_case?